File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/common/uicomponent/gift/giftcloudserver/GiftCloudServer.java
package com.trtc.uikit.livekit.common.uicomponent.gift.giftcloudserver;

import static com.trtc.uikit.livekit.common.uicomponent.gift.giftcloudserver.GiftCloudServerConfig.GIFT_DATA_URL;
import static com.trtc.uikit.livekit.common.uicomponent.gift.giftcloudserver.GiftCloudServerConfig.TE_GIFT_DATA_URL;
import static com.trtc.uikit.livekit.common.uicomponent.gift.giftcloudserver.IGiftCloudServer.Error.BALANCE_INSUFFICIENT;
import static com.trtc.uikit.livekit.common.uicomponent.gift.giftcloudserver.IGiftCloudServer.Error.NO_ERROR;
import static com.trtc.uikit.livekit.common.uicomponent.gift.giftcloudserver.IGiftCloudServer.Error.PARAM_ERROR;

import com.google.gson.Gson;
import com.tencent.qcloud.tuicore.TUICore;
import com.tencent.qcloud.tuicore.TUIThemeManager;
import com.trtc.uikit.livekit.common.uicomponent.gift.model.GiftBean;
import com.trtc.uikit.livekit.common.uicomponent.gift.model.TUIGift;
import com.trtc.uikit.livekit.common.utils.HttpGetRequest;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;

public class GiftCloudServer implements IGiftCloudServer {

    private final String mGiftUrl;

    private int mBalance = 500;

    private final List<TUIGift> mCacheGiftList = new ArrayList<>();

    public GiftCloudServer() {
        mGiftUrl = TUICore.getService("TUIEffectPlayerService") == null ? GIFT_DATA_URL : TE_GIFT_DATA_URL;
    }

    @Override
    public void rechargeBalance(Callback<Integer> callback) {
        mBalance += 100;
        if (callback != null) {
            callback.onResult(NO_ERROR, mBalance);
        }
    }

    @Override
    public void queryBalance(Callback<Integer> callback) {
        if (callback != null) {
            callback.onResult(NO_ERROR, mBalance);
        }
    }

    @Override
    public void queryGiftInfoList(Callback<List<TUIGift>> callback) {
        if (mCacheGiftList.isEmpty()) {
            queryGiftList((error, result) -> {
                if (error == 0) {
                    synchronized (GiftCloudServer.this) {
                        if (!mCacheGiftList.isEmpty()) {
                            mCacheGiftList.clear();
                        }
                        mCacheGiftList.addAll(result);
                    }
                }
                if (callback != null) {
                    callback.onResult(error, result);
                }
            });
        } else {
            if (callback != null) {
                callback.onResult(NO_ERROR, mCacheGiftList);
            }
        }
    }

    @Override
    public void sendGift(String sender, String receiver, TUIGift gift, int giftCount,
                         Callback<Integer> callback) {
        if (gift == null || giftCount <= 0) {
            if (callback != null) {
                callback.onResult(PARAM_ERROR, mBalance);
            }
            return;
        }
        int newBalance = mBalance - giftCount * gift.price;
        if (newBalance >= 0) {
            mBalance = newBalance;
            if (callback != null) {
                callback.onResult(NO_ERROR, newBalance);
            }
        } else {
            if (callback != null) {
                callback.onResult(BALANCE_INSUFFICIENT, mBalance);
            }
        }
    }

    private void queryGiftList(Callback<List<TUIGift>> callback) {
        HttpGetRequest request = new HttpGetRequest(mGiftUrl, new HttpGetRequest.HttpListener() {
            @Override
            public void onSuccess(String response) {
                Gson gson = new Gson();
                GiftBean giftBean = gson.fromJson(response, GiftBean.class);
                final List<TUIGift> giftDataList = transformGiftInfoList(giftBean);
                if (giftDataList != null) {
                    if (callback != null) {
                        callback.onResult(NO_ERROR, giftDataList);
                    }
                } else {
                    onFailed("query gift list failed!");
                }
            }

            @Override
            public void onFailed(String message) {
                if (callback != null) {
                    callback.onResult(IGiftCloudServer.Error.OPERATION_FAILED, Collections.EMPTY_LIST);
                }
            }
        });
        request.execute();
    }

    private List<TUIGift> transformGiftInfoList(GiftBean giftBean) {
        if (giftBean == null) {
            return null;
        }
        List<GiftBean.GiftListBean> giftBeanList = giftBean.getGiftList();
        if (giftBeanList == null) {
            return null;
        }
        boolean isEnglish = Locale.ENGLISH.getLanguage().equals(TUIThemeManager.getInstance().getCurrentLanguage());
        List<TUIGift> giftInfoList = new ArrayList<>();
        for (GiftBean.GiftListBean bean : giftBeanList) {
            TUIGift gift = new TUIGift();
            gift.giftId = bean.getGiftId();
            gift.giftName = isEnglish ? bean.getGiftNameEn() : bean.getGiftName();
            gift.imageUrl = bean.getImageUrl();
            gift.animationUrl = bean.getAnimationUrl();
            gift.price = bean.getPrice();
            giftInfoList.add(gift);
        }
        return giftInfoList;
    }
}

File: iOS/TUILiveKit/Sources/Common/Gift/GiftCloudServer/GiftCloudServer.swift
//
//  GiftCloudServer.swift
//  TUILiveKit
//
//  Created by krabyu on 2024/4/3.
//

import Foundation
import TUICore
import RTCCommon

class GiftCloudServer: IGiftCloudServer {
    private var balance = 500

    func rechargeBalance(callback: @escaping (TUIGiftServerError, Int) -> Void) {
        balance += 100
        callback(.noError, balance)
    }

    func queryBalance(callback: @escaping (TUIGiftServerError, Int) -> Void) {
        callback(.noError, balance)
    }

    func queryGiftInfoList(callback: @escaping (TUIGiftServerError, [TUIGift]) -> Void) {
        DispatchQueue.global().async {
            let giftUrl = self.hasTCEffectPlayer() ? TE_GIFT_DATA_URL : GIFT_DATA_URL
            guard let url = URL(string: giftUrl) else { return }
            let session = URLSession.shared
            let task = session.dataTask(with: url) { data, _, error in
                if error != nil {
                    callback(.operationFailed, [])
                    return
                }
                do {
                    guard let data = data else { return }
                    let json = try JSONSerialization.jsonObject(with: data, options: []) as! [String: Any]
                    guard let giftList = json["giftList"] as? [Any] else { return }
                    let gifts = self.initTUIGifts(giftList: giftList)
                    callback(.noError, gifts)
                } catch let error as NSError {
                    print(error.localizedDescription)
                }
            }
            task.resume()
        }
    }

    func sendGift(sender: String, 
                  receiver: String,
                  giftModel: TUIGift,
                  giftCount: Int,
                  callback: @escaping (TUIGiftServerError, Int) -> Void) {
        guard giftCount > 0 else {
            callback(.paramError, balance)
            return
        }

        let newBalance = balance - giftCount * giftModel.price
        if newBalance >= 0 {
            balance = newBalance
            callback(.noError, newBalance)
        } else {
            callback(.balanceInsufficient, balance)
        }
    }
    
    private func initTUIGifts(giftList: [Any]) -> [TUIGift] {
        var giftArray: [TUIGift] = []
        
        for gift in giftList {
            guard let dict = gift as? [String: Any] else { return [] }
            let giftId = (dict["giftId"] as? String) ?? ""
            let imageUrl = (dict["imageUrl"] as? String) ?? ""
            let animationUrl = (dict["animationUrl"] as? String) ?? ""
            let giftName = LocalizedLanguage.isChinese ? ((dict["giftName"] as? String) ?? "") : ((dict["giftName_en"] as? String) ?? "")
            let price = (dict["price"] as? Int) ?? 99
            let model = TUIGift(giftId: giftId, giftName: giftName, imageUrl: imageUrl, animationUrl: animationUrl, price: price, extInfo: [:])
            giftArray.append(model)
        }
        return giftArray
    }
    
    private func hasTCEffectPlayer() -> Bool {
        let service = TUICore.getService("TUIEffectPlayerService")
        return service != nil
    }
}
